### PR TITLE
fix: use first echo data in t2s optcomb for zero voxels

### PR DESCRIPTION
**Pull request by emdupre**
Closes None. 

In response to a bug report from @shotgunosine on NeuroStars, fMRIPrep expects that optimally combined data has no masking applied. Since voxels with no T2* value create a divide by zero error, these were traditionally excluded from the calculation. This fix uses the first echo's T2* value instead of zero as the filler value.

### CODE REVIEW

File: tedana/combine.py [hunk 1]
```diff
@@ -151,10 +154,9 @@ def make_optcom(data, tes, mask, t2s=None, combmode='t2s', verbose=True):
         else:
             msg = ('Optimally combining data with voxel- and volume-wise T2 '
                    'estimates')
-        t2s = t2s[mask, ..., np.newaxis]  # mask out empty voxels/samples
+        t2s = t2s[mask, ..., np.newaxis]  # add new axis
```
Comment (tsalo) [inline tedana/combine.py]: Should drop mask here I think.

File: tedana/combine.py [hunk 2]
```diff
@@ -33,19 +33,23 @@ def _combine_t2s(data, tes, ft2s):
     """
     n_vols = data.shape[-1]
     alpha = tes * np.exp(-tes / ft2s)
```
Comment (tsalo) [inline tedana/combine.py]: Should mask ft2s right before this I think.

**Comment by emdupre** [comment]
OK, these changes don't seem to affect component selection, but they're creating odd boundaries in the optimally combined `ts_OC` files:

Screenshot from 2019-09-11 14-06-49
*Right is with this PR, left is without.*

I'd suggest this because the first echo (which we're using to fill empty voxels, here) will have values in the greatest number of voxels, but it will also have the highest contrast. I'm really not sure how nicely fMRIPrep will play with these kinds of images -- do you want to try pinning to this PR, @shotgunosine, and see how if it affects performance for the registration ?

**Comment by jbteves** [comment]
@emdupre it looks like our unit tests will also need to be revised if we go this route.

**Comment by dowdlelt** [comment]
Why the preference to use the first echo value? Could it be preferred to use some some interpolation of the T2* map - that way image contrast/borders are maintained?  Or would that get even worse in some areas? I

**Comment by jbteves** [comment]
@dowdlelt if you did it that way wouldn't you get dropout in frontal areas? Or am I misunderstanding how this mask was generated?

**Comment by tsalo** [comment]
> Why the preference to use the first echo value? Could it be preferred to use some some interpolation of the T2* map - that way image contrast/borders are maintained? Or would that get even worse in some areas?

@dowdlelt Could you clarify what you mean by "interpolation of the T2* map"? I don't think that the T2* map will have valid values for those voxels.

**Comment by dowdlelt** [comment]
Probably I am misunderstanding something:
This is a case where the data have been optimally combined - but now that optimally combined data isn't "whole" - it has been masked because there are bad T2* values. Is that right? 

What is strange to me is that for the image provided the edges are sharp - what made those superior and anterior sections so cleanly have bad T2star values? Or is that just a fabricated example?

@tsalo I'm saying spatially interpolate the bad T2star values - if there is a hole (negative/zero value), then use the closest X neighbors (>0) to estimate it. I suppose this would need to take place within some brain mask (based on e1?), because we don't want to fill in background. 

Of course, this opens a whole new can of worms for what happens during PCA/ICA, so perhaps a terrible solution. Though I still get the feeling that I am just missing something here.

**Comment by Shotgunosine** [comment]
I tried running that fix inside the 1.4.1 and 1.5.0 fmriprep docker containers and either I didn't successfully get the new code to run or the change didn't impact anything. Here's what I ran for 1.4.1:
```
docker run -ti --rm -v /EDB/MBDU/nielsond/multi_echo_testing/multiecho_masking_test/:/data:ro -v /EDB/MBDU/nielsond/multi_echo_testing/test_out:/out -v  /EDB/MBDU/nielsond/multi_echo_testing/fs_license:/license --entrypoint "/bin/bash" poldracklab/fmriprep:1.4.1 

cd /src; git clone https://github.com/ME-ICA/tedana; cd tedana/; git fetch origin pull/393/head:mask_fix; git checkout mask_fix; cd ..; pip install -e ./tedana/ ; cd /out; /usr/local/miniconda/bin/fmriprep /data /out/out participant --fs-license-file /license/license.txt --output-spaces MNI152NLin2009cAsym -w /out/wrk
```

and got basically the same output as last time:
Screen Shot 2019-09-24 at 8 03 16 AM

And for 1.5.0:
```
docker run -ti --rm -v /EDB/MBDU/nielsond/multi_echo_testing/multiecho_masking_test/:/data:ro -v /EDB/MBDU/nielsond/multi_echo_testing/test_out:/out -v  /EDB/MBDU/nielsond/multi_echo_testing/fs_license:/license --entrypoint "/bin/bash" poldracklab/fmriprep:1.5.0

cd /src; git clone https://github.com/ME-ICA/tedana; cd tedana/; git fetch origin pull/393/head:mask_fix; git checkout mask_fix; cd ..; pip install -e ./tedana/ ; cd /out; /usr/local/miniconda/bin/fmriprep /data /out/out_15 participant --fs-license-file /license/license.txt --output-spaces MNI152NLin2009cAsym -w /out/wrk_15
```
Screen Shot 2019-09-24 at 8 03 48 AM

Still looks like it's masked when it's not supposed to be and cutting off bits of brain. @emdupre @effigies I'm not sure 100% that what I did will substitute in the updated Tedana code, is there another way I should run this?

**Comment by emdupre** [comment]
Could you pin this commit in the `requirements.txt` and then [re-build the docker image](https://fmriprep.readthedocs.io/en/stable/contributors.html#rebuilding-docker-image) ?

**Comment by Shotgunosine** [comment]
It looks like fmriprep doesn't have a requirements.txt anymore.

**Comment by Shotgunosine** [comment]
So I added:
`tedana @ git+https://github.com/emdupre/tedana.git@fix/t2smap-wf` to `setup.cfg`
 but now I'm getting a conflict when I'm trying to build the docker container:
```
Collecting tedana@ git+https://github.com/emdupre/tedana.git@fix/t2smap-wf from git+https://github.com/emdupre/tedana.git@fix/t2smap-wf (from fmriprep==0.0.0)
  Cloning https://github.com/emdupre/tedana.git (to revision fix/t2smap-wf) to /tmp/pip-install-5f3m943s/tedana
  Running command git clone -q https://github.com/emdupre/tedana.git /tmp/pip-install-5f3m943s/tedana
  Running command git checkout -b fix/t2smap-wf --track origin/fix/t2smap-wf
  error: Your local changes to the following files would be overwritten by checkout:
        docs/make.bat
  Please, commit your changes or stash them before you can switch branches.
  Aborting
ERROR: Command "git checkout -b fix/t2smap-wf --track origin/fix/t2smap-wf" failed with error code 1 in /tmp/pip-install-5f3m943s/tedana
```

**Comment by jbteves** [comment]
@Shotgunosine you don't need the `-b` flag when using `--track`, so try getting rid of `-b fix/t2smap-wf`

**Comment by emdupre** [comment]
> So I added:
tedana @ git+https://github.com/emdupre/tedana.git@fix/t2smap-wf to setup.cfg
but now I'm getting a conflict when I'm trying to build the docker container:

I assume you haven't made any changes to that file ? I've never seen a git error like that when building a docker container.

**Comment by jbteves** [comment]
@emdupre I don't have much experience with `docker`, but is there a way he could insert `git status` for us to debug further?

**Comment by Shotgunosine** [comment]
@jbteves The git command is being generated by the pip install processing the the setup.cfg file. I'm not adding that -b at all.
When I just run git clone -q https://github.com/emdupre/tedana.git.
It lists docs/make.bat as a modified file. Is there a stashed chance or something like that?

**Comment by Shotgunosine** [comment]
My guess is it's related to this: https://confluence.atlassian.com/bitbucketserverkb/git-status-shows-as-files-modified-directly-after-clones-943969037.html

but I'm unsure how to resolve that inside the docker container.

**Comment by emdupre** [comment]
Ooof. I'm also unsure. Are you on a Mac or Windows, @Shotgunosine ? 

Is this still the right dataset to test on, if so ? https://drive.google.com/file/d/1-FkTf06ln7JWElUdS4Nco8t1C4HaMSr6/view

**Comment by Shotgunosine** [comment]
Both myself and the container are on linux. @emdupre can you commit to that branch from a linux box?

**Comment by Shotgunosine** [comment]
Yep, that's still what I'm trying to test on.

**Comment by emdupre** [comment]
> Both myself and the container are on linux. @emdupre can you commit to that branch from a linux box? 

I'm also on a linux box. Trying to replicate on just one run of that data....

**Comment by emdupre** [comment]
OK, got the error. The fix is just to install from the tarball in `setup.cfg`:

`tedana @ https://github.com/emdupre/tedana/tarball/fix/t2smap-wf`

The build (`docker build -t fmriprep-t2s .`) completes successfully. I've then tried running it with 

```
docker run -it -v /home/emdupre/Desktop/dylan_data:/data:ro \
-v /home/emdupre/Desktop/tmp:/out \
-v /home/emdupre/Downloads:/license:ro \
fmriprep-t2s /data /out participant --t2s-coreg \
--fs-license-file /license/license.txt
```

and I'm getting a versioning error:

```
Traceback (most recent call last):
  File "/usr/local/miniconda/bin/fmriprep", line 10, in <module>
    sys.exit(main())
  File "/usr/local/miniconda/lib/python3.7/site-packages/fmriprep/cli/run.py", line 311, in main
    opts = get_parser().parse_args()
  File "/usr/local/miniconda/lib/python3.7/site-packages/fmriprep/cli/run.py", line 49, in get_parser
    currentv = Version(__version__)
  File "/usr/local/miniconda/lib/python3.7/site-packages/packaging/version.py", line 221, in __init__
    raise InvalidVersion("Invalid version: '{0}'".format(version))
packaging.version.InvalidVersion: Invalid version: ''
```

This is still listed as a [recommended development process](https://fmriprep.readthedocs.io/en/stable/contributors.html#rebuilding-docker-image) on the fMRIPrep docs, so I'm not sure why it would be failing ? Would be curious if you see the same error, @Shotgunosine !

**Comment by Shotgunosine** [comment]
I was able to get to at least begin running by passing in a `VERSION` variable in the docker build command:
`docker build --build-arg VERSION=1.5.1rc1 -t fmriprep_local .`

I'll let you know how it turns out.

**Comment by Shotgunosine** [comment]
Running off of master with that change to the setup.py cfg failed at some point, but it ran long enough to produce the boldref and brain mask and, as you can see, the brain is till masked in this step and the brain mask is messed up. @emdupre did you get the same result?
Screen Shot 2019-09-26 at 11 53 07 AM

These are the two commands I ran:
```
docker build --build-arg VERSION=1.5.1rc1 -t fmriprep_local .
docker run -ti --rm -v /EDB/MBDU/nielsond/multi_echo_testing/multiecho_masking_test/:/data:ro -v /EDB/MBDU/nielsond/multi_echo_testing/test_out:/out -v  /EDB/MBDU/nielsond/multi_echo_testing/fs_license:/license  fmriprep_local /data /out participant --t2s-coreg --fs-license-file /license/license.txt
```

And here's my setup.cfg file for fmriprep:
```
install_requires =
    niworkflows ~= 0.10.3
    smriprep ~= 0.3.2
    templateflow ~= 0.4.1
    nibabel >=2.2.1
    indexed_gzip >=0.8.8
    nilearn !=0.5.0, !=0.5.1
    nipype >=1.2.2
    psutil >=5.4
    pybids ~= 0.9.3
    tedana @ https://github.com/emdupre/tedana/tarball/fix/t2smap-wf
    nitime
    numpy
    pandas
    pyyaml
    scikit-image
    sdcflows ~= 0.1.0
    statsmodels
test_requires =
    coverage
    codecov
    pytest
```

**Comment by emdupre** [comment]
It's unfortunate this didn't work.... I'll go ahead and close this, since it seems it won't be the fix we need, but I'll follow-up on https://github.com/poldracklab/fmriprep/issues/1786

**Pull request closed without merging**